### ISSUE TITLE
GTK 3.21 regression: custom background problems with transparent panel with in-process applets

### ISSUE THREAD

**Issue by lukefromdc** [mate-desktop/mate-panel#468]
Somewhere between July 19 and July 22, a change to GTK 3.21.4 caused application of custom color backgrounds to fail on the clock button, tray, and show desktop button. Also corruption of window list button display until panel restarted. Solid color and image backgrounds still work, but the panel must be restarted to fully apply this.

Looks like 
https://git.gnome.org/browse/gtk+/commit/?id=2c7b21718f18ab743cd79784ae28cad582f58503
again, as my June 9 build tested OK but not the June 10 build-and all my "cajafix" builds from before we fixed the desktop also gave perfect custom background application.
##

**Comment by lukefromdc** [comment]
Found a workaround: For GTK 3.21, build all applets out of process. Only in-process applets are affected, so building for GTK 3.22 with all applets out of process fixes this. In-process applets get redraw problems when custom backgrounds are used with GTK 3.21 so distros will need to know never to build applets in-process with this GTK version. Didn't test the fish but the clock, tray, and wncklet components were all affected, and I had them all in-process. Rebuilt with them out of process, no more rendering issues with any background or background change.

**Comment by raveit65** [comment]
Looks like the issue is landed in fedora 25, but abrt will give me stacktraces.
I will attach them later.

**Comment by raveit65** [comment]
Hmm, using '--with-in-process-applets=none' seems to fix the issue for applets in mate-panel package.
I noticed that system-monitor applet is also affected, of course it crashes very often.
But man pages told me that using 'none' is the default.
```
  --with-in-process-applets=APPLETS
                          comma-separated list of applets to compile
                          in-process (possible values: none, clock, fish,
                          notification-area, wncklet, all) [default=none]
```
Looks like the default does not work.

**Comment by lukefromdc** [comment]
Building with no applets at all in-process is what stopped the rendering
bugs for me. Just added system monitor applet to my panel to test for 
a while, so far seeing no issues but by git master version of GTK3 is a
bit more recent too,.

**Comment by monsta** [comment]
I wonder if it affects gnome-panel - IIRC it has in-process applets by default these days.

**Comment by lukefromdc** [comment]
Gnome-panel has problems too-but I was able to fix them entirely in
my theme. I will not try a build of mate-panel with in-process applets
as see if it works there too. 

For gnome-panel, gnome-panel-menu-bar had to be explicltly set transparent
for GTK 3.21 and panel-toplevel.background.horizonatal  used to set the 
background color or image to permit custom backgrounds to work,

**Comment by lukefromdc** [comment]
I just did a little more experimentation. Unlike gnome-panel I had trouble 
in all themes if applets were built in-process, and on changing backgrounds
or using any kind of transparency within the panel (as on my window button 
css designed to work with custom backgrounds) the same kind of redraw bug
that spoiled the desktop in Caja 1.14 with GTK 3.21 was evident. It affected
redrawing after a background type change and in my theme messed up the taskbar
buttons due to use of transparency.

**Comment by raveit65** [comment]
@monsta 
i just tested 1.17.0 tarball building w/o --with-in-process-applets=none option and using a custom background with an alpha value in f25 (gtk3-3.22.2-2.fc25.x86_64)
I see no redraw issue in applet buttons and i can remove/add them w/o a crash or abrt noticed a crash in the background.
But i will test this more.
Anyway, i will use this option for official fedora builds, it doesn't hurts....

**Comment by monsta** [comment]
It won't hurt, I just wanted to make sure that using "none" means the same as not using the option at all.

**Comment by lukefromdc** [comment]
Wonder if any of the changes involving avoiding deprecated gdk_window_set_background functions 
from
https://github.com/mate-desktop/mate-panel/issues/468
provide any clues for this one?

**Comment by lukefromdc** [comment]
gdk_window_set_background_pattern does not seem to be the problem here. An experimental build with every instance of it removed broke custom background drawing as expected, but had to effect on this redraw bug, which seems to be a FOREGROUND redraw issue. Since this is only when applets are built in-process, I'm guessing this is from drawing directly on the root window, which GTK 3.22 does not like. An out of process applet presumably draws to a separate GtkPlug window and thus is not affected.

**Comment by raveit65** [comment]
Isn't this report not the same as https://github.com/mate-desktop/mate-panel/issues/499 at least ?

**Comment by lukefromdc** [comment]
This is a redrawing failure, not a crash. Affected applets are in-process applets only, mate-panel builds by default with all applets out of process. Looks like the foreground is getting picked up with the panel background copied for drawing the applet background, just like what we went through with Caja.

**Comment by lukefromdc** [comment]
Branch https://github.com/mate-desktop/mate-panel/tree/dev-inprocess-wip is a new branch that removes the PanelPlug object from in-process applets, thus fixing their rendering with GTK 3.22. It was a long list of GNOME panel commits to get it working, and I've got one remaining issue I have yet to find a fix for: some kind of introspection issue keeps the invest applet from finding the applet factory, yielding this error:
```

  File "/usr/libexec/invest-applet", line 99, in <module>
    MatePanelApplet.Applet.factory_main(
AttributeError: type object 'Applet' has no attribute 'factory_main'
```
If MatePanelApplet-4.0.typelib from this build is replaced with one from a build of master after installation, invest then works with no other changes. Something is keeping this file from including the factory but I have yet to find it. I don't know much about how introspection works yet. 

All the other applets are working, backgrounds render fine both in and out of process, and a test with GTK 3.16 also worked. One final issue is that an out of process applet dragged from a horizontal to a vertical panel, then back may render transparent or keep its .vertical style class until a panel restart.

**Comment by lukefromdc** [comment]
Found a fix for the invest applet: adding
`from os.path import abspath, dirname, exists, isdir, isfile, join`
as the second line of the invest-applet executable (python) file gets the applet working

**Comment by lukefromdc** [comment]
Thought I had a fix for invest, but had the wrong panel version installed. 
`from os.path import abspath, dirname, exists, isdir, isfile, join` 
had no effect, same failure to find the factory

**Comment by lukefromdc** [comment]
Updated https://github.com/mate-desktop/mate-panel/tree/dev-in-process-bgfix is cleaned up a bit with an interactive rebase and a some squashed commits, though the changes to the commit messages (libpanel-applet to libmate-panel-applet) didn't take for some reason.

**Comment by muktupavels** [comment]
Maybe `factory_main` is not included in introspection? Check `mate_panel_applet_factory_main` function comment. You renamed function name in comment to `panel_applet_factory_main`... That function does not exists, most likely there is warning or error when you build introspection.

**Comment by lukefromdc** [comment]
Thanks! That was it, fixing it let the invest applet work

**Comment by lukefromdc** [comment]
Some change in Debian Unstable a couple weeks ago put a stop to the redraw problems. A test with a tarball from June 7 built with in-process applets verified this was not from a change in mate-panel, and rolling back GTK3 to 3.22.16 showed that was not the change either. Same for a glib rollback to 2.52.1, again no redraw issue. This, some underlying library change in Debian Unstable fixed this, but no idea if this fixes a build on Debian Stable, and no idea if this applies to any other distro.

https://github.com/mate-desktop/mate-panel/pull/622 remains relevant both because whatever got fixed might not show up everywhere, and because removing the GtkPlugs from in-process applets and building with ALL applets in-process will be necessary for any future transition to Wayland. This is because GtkPlug and GtkSocket are X11 only. That in turn is because they use Xembed.

**Comment by lukefromdc** [comment]
https://github.com/mate-desktop/mate-panel/issues/635

Appears to be a subcase of this, as it can also be seen on the window-list applet on a distro  where it still occurs such as both Mageia 5 and Mageia 6.

**Comment by lukefromdc** [comment]
This issue has returned in Debian Unstable after the Aug 2 round up updates, as mysteriously as it disappeared. There were gcc updates included, so I wonder if compiler changes are varying the built binary just enough to make the redraw issue come and go? Tested rolling back libc6, no difference there. Too many changes to test them all with updates about once a week(when I can get landline bandwidth to pull them).

**Comment by lukefromdc** [comment]
The mystery of why this ever seemed resolved in Debian is now solved: I am using a system background that is the same as the custom background uses for testing-and had simply forgotten that only the custom background has this problem and system backgrounds always work.

**Comment by lukefromdc** [comment]
FIXED as of the mate-panel 1.19.3 release be a range of commits
https://github.com/mate-desktop/mate-panel/commit/025a11fdaa76111afc8e03faf6ac36b17028e585
through
https://github.com/mate-desktop/mate-panel/commit/9f43dd5c7c37b84960d06a98a4738ee40d8bfd50
which remove the GtkPlugs from in-process applets, turning them transparent over the panel
